### PR TITLE
attempt at a workaround for issues determining clock skew on k8s 1.7

### LINKED ISSUE #3943: Clock skew check not working correctly
### What is the issue?
Linkerd --pre (randomly) fails on clock skew detection check

### How can it be reproduced?
If I were to guess why, use Kubernetes 17.0.0, made sure to check ntpd many times, also switched back and forth between ntpd and systemd-timesyncd

./clock_skew.sh is a lightly modified version of this script:
https://gist.github.com/foospidy/77c43ec2b2b71e4447f0534f7ac26798

### Logs, error output, etc
'proof'
```
/clock_skew.sh 
INFO: Clock skew on 0.0.0.27 is 0 seconds.
INFO: Clock skew on 0.0.0.51 is 1 seconds.
INFO: Clock skew on 0.0.0.161 is 0 seconds.
INFO: Clock skew on 0.0.0.113 is 0 seconds.
Everything is running on time!
```
simple capturing of node metadata
```
kubectl describe nodes | grep Ready | grep posting
  Ready                True    Fri, 17 Jan 2020 14:15:39 +0100   Fri, 17 Jan 2020 13:45:28 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:15:28 +0100   Fri, 17 Jan 2020 13:45:27 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:15:28 +0100   Fri, 17 Jan 2020 13:45:27 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:15:28 +0100   Fri, 17 Jan 2020 13:45:27 +0100   KubeletReady                 kubelet is posting ready status
```

#### EDIT
Whilst verifying output for this bug report I saw a massive jump in heartbeat output
```text
 kubectl describe nodes | grep Ready | grep posting
  Ready                True    Fri, 17 Jan 2020 14:30:40 +0100   Fri, 17 Jan 2020 13:45:28 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:37:17 +0100   Fri, 17 Jan 2020 13:45:27 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:30:28 +0100   Fri, 17 Jan 2020 13:45:27 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:30:28 +0100   Fri, 17 Jan 2020 13:45:27 +0100   KubeletReady                 kubelet is posting ready status
./clock_skew.sh 
INFO: Fri, 17 Jan 2020 14:37:53 +0100 vs Fri, 17 Jan 2020 13:37:54 +0000 Clock skew on 0.0.100.27 is 1.
INFO: Fri, 17 Jan 2020 14:37:53 +0100 vs Fri, 17 Jan 2020 13:37:54 +0000 Clock skew on 0.0.101.51 is 1.
INFO: Fri, 17 Jan 2020 14:37:54 +0100 vs Fri, 17 Jan 2020 13:37:54 +0000 Clock skew on 0.0.100.161 is 0.
INFO: Fri, 17 Jan 2020 14:37:54 +0100 vs Fri, 17 Jan 2020 13:37:54 +0000 Clock skew on 0.0.100.113 is 0.
Everything is running on time!
kubectl describe nodes | grep Ready | grep posting
  Ready                True    Fri, 17 Jan 2020 14:37:24 +0100   Fri, 17 Jan 2020 13:45:28 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:37:17 +0100   Fri, 17 Jan 2020 13:45:27 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:37:23 +0100   Fri, 17 Jan 2020 14:37:23 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:37:23 +0100   Fri, 17 Jan 2020 13:45:27 +0100   KubeletReady                 kubelet is posting ready status
./clock_skew.sh                                   
INFO: Fri, 17 Jan 2020 14:38:13 +0100 vs Fri, 17 Jan 2020 13:38:14 +0000 Clock skew on 0.0.100.27 is 1.
INFO: Fri, 17 Jan 2020 14:38:13 +0100 vs Fri, 17 Jan 2020 13:38:13 +0000 Clock skew on 0.0.101.51 is 0.
INFO: Fri, 17 Jan 2020 14:38:14 +0100 vs Fri, 17 Jan 2020 13:38:13 +0000 Clock skew on 0.0.100.161 is -1.
INFO: Fri, 17 Jan 2020 14:38:14 +0100 vs Fri, 17 Jan 2020 13:38:14 +0000 Clock skew on 0.0.100.113 is 0.
Everything is running on time
```

And whilst investigating _that_ output it started lagging again
```
kubectl describe nodes | grep Ready | grep posting
  Ready                True    Fri, 17 Jan 2020 14:37:24 +0100   Fri, 17 Jan 2020 13:45:28 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:37:17 +0100   Fri, 17 Jan 2020 13:45:27 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:37:23 +0100   Fri, 17 Jan 2020 14:37:23 +0100   KubeletReady                 kubelet is posting ready status
  Ready                True    Fri, 17 Jan 2020 14:37:23 +0100   Fri, 17 Jan 2020 13:45:27 +0100   KubeletReady                 kubelet is posting ready status
./clock_skew.sh                                   
INFO: Fri, 17 Jan 2020 14:38:57 +0100 vs Fri, 17 Jan 2020 13:38:58 +0000 Clock skew on 0.0.100.27 is 1.
INFO: Fri, 17 Jan 2020 14:38:57 +0100 vs Fri, 17 Jan 2020 13:38:58 +0000 Clock skew on 0.0.101.51 is 1.
INFO: Fri, 17 Jan 2020 14:38:58 +0100 vs Fri, 17 Jan 2020 13:38:58 +0000 Clock skew on 0.0.100.161 is 0.
INFO: Fri, 17 Jan 2020 14:38:58 +0100 vs Fri, 17 Jan 2020 13:38:58 +0000 Clock skew on 0.0.100.113 is 0.
Everything is running on time
```

I'm using bootkube and couldn't find any override for the default heartbeat settings  is it possible the heartbeat output simply isn't reliable enough (any more) ?

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

pre-kubernetes-setup
--------------------
√ control plane namespace does not already exist
√ can create Namespaces
√ can create ClusterRoles
√ can create ClusterRoleBindings
√ can create CustomResourceDefinitions
√ can create PodSecurityPolicies
√ can create ServiceAccounts
√ can create Services
√ can create Deployments
√ can create CronJobs
√ can create ConfigMaps
× no clock skew detected
    clock skew detected for node(s): supernova-controller-0, supernova-helium-worker-0, supernova-helium-worker-1, supernova-helium-worker-2
    see https://linkerd.io/checks/#pre-k8s-clock-skew for hints

pre-kubernetes-capability
-------------------------
√ has NET_ADMIN capability
√ has NET_RAW capability

pre-linkerd-global-resources
----------------------------
√ no ClusterRoles exist
√ no ClusterRoleBindings exist
√ no CustomResourceDefinitions exist
√ no MutatingWebhookConfigurations exist
√ no ValidatingWebhookConfigurations exist
√ no PodSecurityPolicies exist

linkerd-version
---------------
√ can determine the latest version
‼ cli is up-to-date
    is running version 2.6.0 but the latest stable version is 2.6.1
    see https://linkerd.io/checks/#l5d-version-cli for hints
```

### Environment

- Kubernetes Version: 1.17.0
- Cluster Environment: Packet (Bootkube)
- Host OS: Flatcar Linux
- Linkerd version: 2.6.1

### Possible solution
Checked the code (pkg/healthcheck/healthcheck.go), nothing stood out to me as to a possible cause of this. But maybe making it commandline overrideable would be nice. Now I have to ignore the check output in my terraform scripts instead

### PR DESCRIPTION
**Pull request by DGollings**
Subject

Problem
Clock skew not measurable in 1.7

Solution
use nodes themselves as a form of clock quorum

Validation
Added some tests

Fixes  #3943

### CODE REVIEW

**Comment by DGollings** [comment]
ps. at some point whilst figuring out to get the DCO check to pass I accidentally 'added' another commit?

File: charts/linkerd2-cni/README.md [hunk 1]
```diff
@@ -13,7 +13,7 @@ The following table lists the configurable parameters of the Linkerd2-cni chart
 
 | Parameter                            | Description                                                           | Default                       |
 |--------------------------------------|-----------------------------------------------------------------------|-------------------------------|
-`cniPluginImage`                      | Docker image for the CNI plugin                                       |`gcr.io/linkerd-io/cni-plugin`|
```
Comment (olix0r) [inline charts/linkerd2-cni/README.md]: It looks like master probably changed under you since you branched. Can you merge master into your branch when you get a chance to avoid these unrelated changes?

**Comment by DGollings** [comment]
Sure, will do so tomorrow. How do you feel about the health check code itself?

**Comment by olix0r** [comment]
It took me a minute to find the relevant change.. I think this is most of it: https://github.com/linkerd/linkerd2/pull/3945/files#diff-d4056ff163bcf2aeacefb2a34164563cR1842-R1896

I like the way this change collects the range of skews to compare; I think this is promising. We'll want to get a few more eyes on this once the diff is a bit narrower, though. Thanks for putting this together!
